### PR TITLE
Projectile refactoring 3

### PR DESCRIPTION
**Pull request by MrRowey**
Batch 3 of projectile Refactoring

### CODE REVIEW

**Review by Garanas — CHANGES_REQUESTED** [review]
As with the other two pull requests - looking good. But carefully evaluate the `self:ForkThread` and `WaitSeconds` replacements

File: projectiles/CDFTracker01/CDFTracker01_script.lua [hunk 1]
```diff
@@ -7,11 +7,10 @@ CDFTracker01 = Class(CDFTrackerProjectile) {
     OnImpact = function(self, TargetType, TargetEntity)
         if TargetEntity then
             local x,y,z = unpack(TargetEntity:GetPosition())
-            local tracker = CreateUnit('URB5206', self.Data.Parent:GetArmy(), x, y, z, 0, 0, 0, 0)
+            local tracker = CreateUnit('URB5206', self.Data.Parent.Army, x, y, z, 0, 0, 0, 0)
```
Comment (Garanas) [inline projectiles/CDFTracker01/CDFTracker01_script.lua:10]: What is this 'Parent' value? Is that a unit?

File: projectiles/CIFBrackmanHackPegs01/CIFBrackmanHackPegs01_script.lua [hunk 2]
```diff
@@ -15,9 +15,9 @@ CIFBrackmanHackPegs01 = Class(CDFBrackmanHackPegProjectile01) {
         local ChildProjectileBP = '/projectiles/CIFBrackmanHackPegs02/CIFBrackmanHackPegs02_proj.bp'  
         ------ Play split effects
         for k, v in FxFragEffect do
-            CreateEmitterAtEntity( self, self:GetArmy(), v )
+            CreateEmitterAtEntity( self, self.Army, v )
         end
-        local vx, vz = self:GetVelocity()
+        local vx, _, vz = self:GetVelocity()
```
Comment (Garanas) [inline projectiles/CIFBrackmanHackPegs01/CIFBrackmanHackPegs01_script.lua:20]: I'm not sure if this is logically equivalent, but this unit isn't being used regardless 😄

File: projectiles/CIFBrackmanHackPegs02/CIFBrackmanHackPegs02_script.lua [hunk 3]
```diff
@@ -34,8 +34,8 @@ CIFBrackmanHackPegs02 = Class(import("/lua/cybranprojectiles.lua").CDFBrackmanHa
         }
         local num_projectiles= 50
         for i = 0, num_projectiles do
-            WaitTime= ( 0.4+RandomFloat(-0.2,0.2) )  ------Get how long to wait before launching the next one.
-            WaitSeconds( WaitTime )
+            WaitTime= ( 0.4+RandomFloat(-0.2,0.2)*10+1 )  ------Get how long to wait before launching the next one.
```
Comment (Garanas) [inline projectiles/CIFBrackmanHackPegs02/CIFBrackmanHackPegs02_script.lua:37]: This change isn't equivalent, I think it should be: `3 + ((Random() * 3) ^0)`, where `^0` is a cheap way to floor a positive number using bit operators 😄

File: projectiles/CIFEMPFluxWarhead03/CIFEMPFluxWarhead03_script.lua [hunk 4]
```diff
@@ -5,10 +5,5 @@ local EmitterProjectile = import("/lua/sim/defaultprojectiles.lua").EmitterProje
 
 CIFEMPFluxWarhead03 = Class(EmitterProjectile) {
     FxTrails = {'/effects/emitters/empfluxwarhead_04_emit.bp',},
-    FxImpactUnit = {},
```
Comment (Garanas) [inline projectiles/CIFEMPFluxWarhead03/CIFEMPFluxWarhead03_script.lua:8]: These may have use: they can override effects provided in the hierarchy. Are you sure this is a safe change?

File: projectiles/CIFEMPFluxWarhead04/CIFEMPFluxWarhead04_script.lua [hunk 5]
```diff
@@ -6,8 +6,6 @@ local CEMPFluxWarheadProjectile = import("/lua/cybranprojectiles.lua").CEMPFluxW
 CIFEMPFluxWarhead04 = Class(CEMPFluxWarheadProjectile) {
     BeamName = '/effects/emitters/missile_exhaust_fire_beam_06_emit.bp',
     FxSplashScale = 0.5,
-    FxTrails = {},
```
Comment (Garanas) [inline projectiles/CIFEMPFluxWarhead04/CIFEMPFluxWarhead04_script.lua:9]: Same here, see last conversation

File: projectiles/CIFMissileTactical01/CIFMissileTactical01_Script.lua [hunk 6]
```diff
@@ -55,20 +55,19 @@ CIFMissileTactical01 = Class(CLOATacticalMissileProjectile) {
...
         end
-        
+
         CLOATacticalMissileProjectile.OnImpact(self, targetType, targetEntity)
     end,
-    
+
     OnDamage = function(self, instigator, amount, vector, damageType)
-        if not self.Split and (amount >= self:GetHealth()) then
+        if not self.Split and (amount >= self.Health) then
```
Comment (Garanas) [inline projectiles/CIFMissileTactical01/CIFMissileTactical01_Script.lua:70]: self.Health is not populated, best to use the C function

File: projectiles/CIFMissileTactical02/CIFMissileTactical02_Script.lua [hunk 7]
```diff
@@ -16,22 +16,22 @@ CIFMissileTactical02 = Class(CLOATacticalMissileProjectile) {
...
         end
-        
+
         CLOATacticalMissileProjectile.OnImpact(self, targetType, targetEntity)
     end,
-    
+
     OnDamage = function(self, instigator, amount, vector, damageType)
-        if not self.Split and (amount >= self:GetHealth()) then
+        if not self.Split and (amount >= self.Health) then
```
Comment (Garanas) [inline projectiles/CIFMissileTactical02/CIFMissileTactical02_Script.lua:34]: Same here, not populated

File: projectiles/CIFMissileTactical03/CIFMissileTactical03_Script.lua [hunk 8]
```diff
@@ -7,30 +7,27 @@
...
         end
-        
         CLOATacticalMissileProjectile.OnImpact(self, targetType, targetEntity)
-    end,   
-    
+    end,
+
     OnDamage = function(self, instigator, amount, vector, damageType)
-        if not self.Split and (amount >= self:GetHealth()) then
+        if not self.Split and (amount >= self.Health) then
```
Comment (Garanas) [inline projectiles/CIFMissileTactical03/CIFMissileTactical03_Script.lua:30]: Same here, not poulated

**Review by Garanas — CHANGES_REQUESTED** [review]
Make sure to evaluate the hierarchy for certain changes. Removing empty tables that represent effects may have side effects.

Also: `self.Health` is not a field, unless it is defined somewhere and I'm not aware of it 😄

File: projectiles/CDFTracker01/CDFTracker01_script.lua (hunk 1 above)
Comment (MrRowey) [reply projectiles/CDFTracker01/CDFTracker01_script.lua:10]: not sure :)

File: projectiles/CIFEMPFluxWarhead03/CIFEMPFluxWarhead03_script.lua (hunk 4 above)
Comment (MrRowey) [reply projectiles/CIFEMPFluxWarhead03/CIFEMPFluxWarhead03_script.lua:8]: how to do i find if it has a uses as in the hierarchy FxImpactUnit is not populated

File: projectiles/CIFEMPFluxWarhead04/CIFEMPFluxWarhead04_script.lua (hunk 5 above)
Comment (MrRowey) [reply projectiles/CIFEMPFluxWarhead04/CIFEMPFluxWarhead04_script.lua:9]: Same as Last comment

File: projectiles/CIFEMPFluxWarhead03/CIFEMPFluxWarhead03_script.lua (hunk 4 above)
Comment (Garanas) [reply projectiles/CIFEMPFluxWarhead03/CIFEMPFluxWarhead03_script.lua:8]: You're correct, they're all empty

File: projectiles/CIFEMPFluxWarhead04/CIFEMPFluxWarhead04_script.lua (hunk 5 above)
Comment (Garanas) [reply projectiles/CIFEMPFluxWarhead04/CIFEMPFluxWarhead04_script.lua:9]:
```lua
---@class CEMPFluxWarheadProjectile : NukeProjectile
CEMPFluxWarheadProjectile = Class(NukeProjectile, SingleBeamProjectile) {
    BeamName = '/effects/emitters/missile_exhaust_fire_beam_01_emit.bp',
    FxInitialAtEntityEmitter = {},
    FxUnderWaterTrail = {'/effects/emitters/missile_cruise_munition_underwater_trail_01_emit.bp',},
    FxOnEntityEmitter = {},
    FxExitWaterEmitter = EffectTemplate.DefaultProjectileWaterImpact,
    FxSplashScale = 0.65,
    ExitWaterTicks = 9,
    FxTrailOffset = -0.5,

    -- LAUNCH TRAILS
    FxLaunchTrails = {},

    -- TRAILS
    FxTrails = {'/effects/emitters/missile_cruise_munition_trail_01_emit.bp',},

    -- Hit Effects
    FxImpactUnit = {},
    FxImpactLand = {},
    FxImpactUnderWater = {},
}
```

The class it inherits from has fx trails set, so if we'd not set them to an empty table then we'd end up having trails

**Review by Garanas — APPROVED** [review]

File: projectiles/CDFTracker01/CDFTracker01_script.lua (hunk 1 above)
Comment (Garanas) [reply projectiles/CDFTracker01/CDFTracker01_script.lua:10]: this is a weird script, as a projectile ends up creating a unit 🤔 ?

**Pull request merged**